news_app/lib/widgets/news_future_builder.dart

```diff
@@ -0,0 +1,57 @@
+import 'package:flutter/material.dart';
+import 'package:news_app/model/article_model.dart';
+import 'package:news_app/service/news_service.dart';
+import 'package:news_app/widgets/grid_view_builder.dart';
+import 'package:news_app/widgets/news_failed.dart';
+
+class NewsFutureBuilder extends StatefulWidget {
+  const NewsFutureBuilder({super.key});
+
+  @override
+  State<NewsFutureBuilder> createState() => _NewsFutureBuilderState();
+}
+
+class _NewsFutureBuilderState extends State<NewsFutureBuilder> {
+  late Future<List<ArticleModel>> _articles;
+
+  Future<void> _onRefresh() async {
+    setState(() {
+      _articles = NewsService().getNews();
+    });
+    await _articles;
+  }
+
+  @override
+  void initState() {
+    super.initState();
+
+    _articles = NewsService().getNews();
+    setState(() {});
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return RefreshIndicator(
+      onRefresh: _onRefresh,
+      child: FutureBuilder(
+        future: _articles,
+        builder: (context, snapshot) {
+          if (snapshot.hasData) {
+            return GridViewBuilder(
+              // adaptive based on the screen size
+              articles: snapshot.data ?? [],
+            );
+          }
+          if (snapshot.hasError) {
+            return NewsFailed(onRefresh: _onRefresh);
+          }
+          return const Center(
+            child: CircularProgressIndicator(
+              color: Colors.black,
+            ),
+          );
+        },
+      ),
+    );
+  }
+}
```

Comment:
Using power of Riverpod by using the Provider for immutable objects, and FutureProvider to handle the api call and its states

```
final dioProvider = Provider<Dio>((_) => Dio(BaseOptions(baseUrl: 'https://newsapi.org')));
final getNewsFutureProvider =
    FutureProvider<List<ArticleModel>>((ref) => NewsService(dio: ref.watch(dioProvider)).getNews());

class NewsFutureBuilder extends ConsumerStatefulWidget {
  const NewsFutureBuilder({super.key});

  @override
  ConsumerState<NewsFutureBuilder> createState() => _NewsFutureBuilderState();
}

class _NewsFutureBuilderState extends ConsumerState<NewsFutureBuilder> {
  Future<void> _onRefresh() async {
    ref.invalidate(getNewsFutureProvider);
  }

  @override
  Widget build(BuildContext context) {
    return RefreshIndicator(
      onRefresh: _onRefresh,
      child: ref.watch(getNewsFutureProvider).when(data: (data) {
        return GridViewBuilder(
          // adaptive based on the screen size
          articles: data,
        );
      }, error: (_, __) {
        return NewsFailed(onRefresh: _onRefresh);
      }, loading: () {
        return const Center(
          child: CircularProgressIndicator(
            color: Colors.black,
          ),
        );
      }),
    );
  }
}
```